### ISSUE TITLE
Custom Label "app" not populating for Some Reports (Trivy Vulnerability)

### ISSUE THREAD

**Issue by sherifkayad** [kyverno/policy-reporter-ui#213]
probably related to #191 

For some reports, the k8s labels are populated perfectly (when the count of the distinct labels > 35 / 40) .. howver, for some others not .. what am I missing?

Checking the POLR CRDs, it seems to me like all reports have the k8s labels already .. could it be a bug related to the number of values? .. in some other UI instance that I have with like only two or three testing apps, the labels are populated successfully.

Unfortunately neither the API nor the UI pods are showing any special logs determining if there's an issue.

![image-1](https://user-images.githubusercontent.com/8362784/212640455-b9361f70-1a20-44bd-b255-91680caf4438.png)

![image-2](https://user-images.githubusercontent.com/8362784/212643250-0311702a-491e-44d7-9bb4-c381789916d5.png)

**Comment by fjogeleit** [comment]
So you mean that not all labels available in the ui but correctly set on the CRDs?

**Comment by sherifkayad** [comment]
Yeah exactly .. It's really weird as it's happending in my case only for one category of reports. The label is there on the report CRD but not visible in the UI. Actually the UI isn't showing any value at all.

Is there a way to debug more what could be the reason for that? Maybe you can simulate more than like 40 reports or so with distinct `app` label values and see if that works for you?

**Comment by fjogeleit** [comment]
Can you open the network tab and have a look if there is any issue in the console or if some requests failing? Sounds like an issue in the API

**Comment by sherifkayad** [comment]
Nothing much .. I got a bunch of `301`s and `200`s .. nothing that might indicate an error .. The Console Tab is completely error free.

![image](https://user-images.githubusercontent.com/8362784/212644056-624a82b7-a0ef-4373-ab4a-de3a05821822.png)

**Comment by sherifkayad** [comment]
just one thing that got me curious .. why do I actually get that many `301`s? is that a standard behavior?

**Comment by fjogeleit** [comment]
The UI backend works mostly as proxy. So the requests to the UI backend are redirected to the actual API. 

Are there any results in the report-labels or report requests?

**Comment by sherifkayad** [comment]
The report payload request returned the report fine .. however, with the data needed to display it in the UI; that's without the labels. Howver, I believe that's the expected behavior.

**Comment by fjogeleit** [comment]
Okay, thanks for your feedback. I will have a look if I find a possible reason.

**Comment by fjogeleit** [comment]
Ah, just one question, did you try to kill the policy-reporter core pod, to trigger a reprocessing of your reports?

**Comment by sherifkayad** [comment]
Just killed the Policy Reporter / Policy Reporter UI pods now .. unfortunately the same result.

**Comment by sherifkayad** [comment]
just one thing I noticed right now: the UI pod is throwing the following error: `2023/01/16 11:18:31 http: proxy error: context canceled` .. not sure if that's related.

**Comment by fjogeleit** [comment]
okay, thanks for this. Maybe the API is to slow in your env and didn't respond. I will check that.

**Comment by fjogeleit** [comment]
Maybe one thing you could test, forward the core app and try to access the API directly and see if it comes to an timeout or issue.

```
http://localhost:8080/v1/namespaced-resources/report-labels?sources=Trivy%20Vulnerability
```

**Comment by sherifkayad** [comment]
@fjogeleit thanks for getting back to me on this. Really weird the API (in just 39 ms) returns an empty list 😢 .. For the Trivy ConfigAudit as well as on my other smaller cluster, the API responds just fine with the correct list of labels. I think we ruled out the possibility that it's a UI issue. Somehow the API isn't able to retrieve the list from the CRDs, don't know why 😭 

![image](https://user-images.githubusercontent.com/8362784/212866770-81a28f9f-de5e-48d7-b04f-e5f14eeb840a.png)

**Comment by sherifkayad** [comment]
let me know if I should close this ticket and open another one for the API (Policy Reporter Project) ..

**Comment by sherifkayad** [comment]
That's 100% correct. That's my case yes.

**Comment by fjogeleit** [comment]
I am started a build of the core app with `v2.12.0-dev`. It has no fix yet but logs the conversion of the labels to json and the other way arround. I want to be sure that there is no misbehavior. It would be great if you could use this version as image.tag `2.12.0-dev` and check the core app logs when accessing the API.

If no logs are shown it would ensure that the values itself a correctly persisted and read.

**Comment by sherifkayad** [comment]
sure thing can do so. Can you tell me when the job pushes? .. I checked that one here https://github.com/kyverno/policy-reporter/actions/runs/3938402839/jobs/6737046086 but seems like it failed.

**Comment by fjogeleit** [comment]
Yep, sorry. I fixed the build and it is done now.

**Comment by sherifkayad** [comment]
will try right away and send you my feedback

**Comment by sherifkayad** [comment]
@fjogeleit unfortunately the same result .. got 200 success in under 40 ms with an empty list and no logs .. Do I need to increase the log level or so?

![image](https://user-images.githubusercontent.com/8362784/212894708-5f5daec5-9f97-4418-a9fd-1c74505abf75.png)

![image](https://user-images.githubusercontent.com/8362784/212894898-1a6092ee-beeb-4434-a0d7-a2323382ca3c.png)

**Comment by fjogeleit** [comment]
no you don't but thats fine. So I know that the labels are persisted and retrieved correctly. So I will check the query next and see if I find an edge case which explains the issue.

**Comment by sherifkayad** [comment]
great! .. Just if that might help you: on the level of the Prometheus metrics (and of course the dashboards), checking e.g. for the metric `` using e.g. http://localhost:8080/metrics gets the correct result with the correct App Label.

```
policy_report_result{app="platform-demo",category="Vulnerability Scan",kind="ReplicaSet",name="platform-demo-6d9b8565f4",namespace="my-namespace-1",policy="CVE-2017-14159",report="trivy-vuln-polr-replicaset-platform-demo-6d9b8565f4",rule="",severity="low",source="Trivy Vulnerability",status="warn",team=""} 2
```

**Comment by fjogeleit** [comment]
Okay, thanks. This is good to know.

**Comment by fjogeleit** [comment]
Hey, one question, what configuration in the  Trivy Operator do you use to apply resource labels to the Trivy Operator Reprot CRDs?

**Comment by sherifkayad** [comment]
hey there .. I am using the Trivy Operator Helm Chart (currently using `v0.10.1`) (https://github.com/aquasecurity/trivy-operator/blob/v0.10.1/deploy/helm/values.yaml) .. To set that I am using the property `trivyOperator.reportResourceLabels` setting it to `team,app`

e.g.:
```yaml
trivyOperator:  
   ....
  # reportResourceLabels comma-separated scanned resource labels which the user wants to include in the Prometheus
  # metrics report. Example: `owner,app`
  reportResourceLabels: "team,app"
```

With that I get something like:
```yaml
apiVersion: aquasecurity.github.io/v1alpha1
kind: VulnerabilityReport
metadata:
  annotations:
    trivy-operator.aquasecurity.github.io/report-ttl: 24h0m0s
  creationTimestamp: '2023-01-19T16:04:58Z'
  generation: 1
  labels:
    app: platform-demo
    resource-spec-hash: 54b7bb76df
    trivy-operator.container.name: platform-demo
    trivy-operator.resource.kind: ReplicaSet
    trivy-operator.resource.name: platform-demo-5579b74cd9
    trivy-operator.resource.namespace: my-namespace
  managedFields:
    ...
  vulnerabilities:
    - class: os-pkgs
      cvss:
        nvd:
          V2Score: 4.3
          V2Vector: AV:N/AC:M/Au:N/C:N/I:P/A:N
          V3Score: 3.7
          V3Vector: CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:L/A:N
      fixedVersion: ''
      installedVersion: 1.8.2.3
      links: []
      packageType: debian
      primaryLink: https://avd.aquasec.com/nvd/cve-2011-3374
      resource: apt
      score: 3.7
      severity: LOW
      target: >-
        167818453886.dkr.ecr.eu-central-1.amazonaws.com/red-kite-cockpit/platform-demo:v2.0.12
        (debian 10.13)
      title: >-
        It was found that apt-key in apt, all versions, do not correctly valid
        ...
      vulnerabilityID: CVE-2011-3374
    .....
    - class: os-pkgs
    ......
    
    - class: lang-pkgs
      cvss:
        ghsa:
          V3Score: 6.5
          V3Vector: CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:U/C:N/I:N/A:H
        nvd:
          V3Score: 6.5
          V3Vector: CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:U/C:N/I:N/A:H
        redhat:
          V3Score: 6.5
          V3Vector: CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:U/C:N/I:N/A:H
      fixedVersion: '1.32'
      installedVersion: '1.29'
      links: []
      packageType: jar
      primaryLink: https://avd.aquasec.com/nvd/cve-2022-41854
      resource: org.yaml:snakeyaml
      score: 6.5
      severity: MEDIUM
      target: Java
      title: 'dev-java/snakeyaml: DoS via stack overflow'
      vulnerabilityID: CVE-2022-41854
```

Note here that I currently have set only the `app` label on the service and not the `team` yet.

**Comment by fjogeleit** [comment]
To debug a bit further I created a PR with 2 new APIs to list the actual Policy Report information, including the persisted labels.

I will prepare an image for it again, it would be great if you could use it, port forward the core app and see if
1) Policy Report information of your Vulnr Reports are persisted correctly
2) Your labels are persisted correctly

I let you know as soon as the image is available.

The APIs are paginated, the API and the query parameters are:

```
http://localhost:8080/v1/policy-reports?namespaces=<namespace>&offset=20&page=1
```

**Comment by sherifkayad** [comment]
sure thing .. let me know and I am happy to test it.

**Comment by sherifkayad** [comment]
btw you mentioned 2 new debugging APIs, one of them would be this one `http://localhost:8080/v1/policy-reports?namespaces=<namespace>&offset=20&page=1` .. what's the other one?

**Comment by fjogeleit** [comment]
The same for cluster policy reports


`v1/cluster-policy-reports`

**Comment by fjogeleit** [comment]
I found one issue in the `v1/namespaced-resources/report-labels` API, not sure if it will fix your issue but I prepared an image tag `2.12.0-dev3`. So you can try it out. If it is still not working - I also added a namespaces query filter to the API, so it would be good if you could try the API for a single or subset of namespaces.

`v1/namespaced-resources/report-labels?namespaces=<namespace1>&nameapaces=<namespace2>`

**Comment by sherifkayad** [comment]
@fjogeleit I just tried the new image .. thanks for that ..

The outcome is now a bit strange .. I currently have two namespaces with reports .. for namespace 1 e.g. `my-namespace` (`http://localhost:8080/v1/namespaced-resources/report-labels?namespaces=my-namespace`) I get results as shown below ..
![pic1](https://user-images.githubusercontent.com/8362784/213877894-6000bf19-a270-4aa5-8c84-f73b85e34c9d.png)


For my other namespace `my-namespace-2` I don't get anything .. nevertheless for that other namespace I have plenty of reports (Vuln and ConfigAudit) present (also confirmed by `http://localhost:8080/v1/policy-reports?namespaces=my-namespace-2&offset=100&page=1`).
![image](https://user-images.githubusercontent.com/8362784/213877851-8386903c-dea3-49e1-97ec-eb65e38b3092.png)

When I do a combined call (`http://localhost:8080/v1/namespaced-resources/report-labels?namespaces=my-namespace&namespaces=my-namespace-2`) I also get an empty list (similar to above).

**Comment by fjogeleit** [comment]
Okay also it is not working when you try to get the labels for this concrete namespaces and the only difference are the amount of reports but the habe basically the same labels and values?

**Comment by sherifkayad** [comment]
No, when I explicitly called the API filtering for that second namespace I got an empty list.

Both namespaces are identical .. Trivy currently is configured to scan only those two. The only difference is obviously the name of the namespace (one is called `x-y` and the other one (not working one) is called `a-b-c` .. the number of reports in the non working namespace is larger, yes .. and of course the label values are unique in each namespace.

**Comment by sherifkayad** [comment]
@fjogeleit I just noticed the second part of the question .. sure thing I can send you these privately .. let me know how and I am happy to do it.

**Comment by fjogeleit** [comment]
Would be very helpful, thanks. You can send it via email to me. [frank.jogeleit@web.de](mailto:frank.jogeleit@web.de)

**Comment by fjogeleit** [comment]
Just in case that it is really related to the amount of reports and results I reworked the schema and query a bit, so no join is necessary and the query should perform in general a bit better.

I created the tag `v2.12.0-dev4` to test it out.

**Comment by sherifkayad** [comment]
I will try that new image out .. in the worst case that it's still not working, I will send you the generated `PolicyReport` objects in that namespace in question

**Comment by fjogeleit** [comment]
thanks a lot

**Comment by sherifkayad** [comment]
@fjogeleit WOHOO that last change worked perfectly! I assume it had to do with the count! ..

![image1](https://user-images.githubusercontent.com/8362784/213912857-b585f414-3605-41ed-8329-5016ae18c723.png)

I had for that namespace in question `113` reports (combined `Vuln` and `ConfigAudit`) .. Also confirmed by `http://localhost:8080/v1/policy-reports?namespaces=my-namespace-2&offset=200&page=1`

Now all API calls are working perfectly!
- `http://localhost:8080/v1/policy-reports?namespaces=my-namespace-2&offset=200&page=1`
- `http://localhost:8080/v1/namespaced-resources/report-labels?namespaces=my-namespace-2`
- `http://localhost:8080/v1/namespaced-resources/report-labels?namespaces=my-namespace&namespaces=my-namespace-2`
- `http://localhost:8080/v1/namespaced-resources/report-labels`

I also checked the UI and it works like a charm!

Thanks a lot for your really deep and detailed help on that one! Much appreciated.

Is there a way we can know the breaking limit of this? .. It's of course great that it now works .. but let's say tomorrow in one namespace I get 300 or even more reports, how can we ensure that this won't be an issue? .. Especially that the response times of the API were like under `40ms` ..

I will also leave it up to you when this ticket can be closed (maybe you need to do some clean-up work on the APIs or so).

**Comment by fjogeleit** [comment]
Great news, really happy that it is working now. I can not really say if it was related to the amount of reports / results or a special case with the old query.

The last version joined the policy report table with the results table. Because each report can have a lot of results, this Influenced the query a lot. Now that the policy-report table has all needed information the JOIN is no longer needed and the query should scale much better. I don't think that 200 reports or more are a problem now - even though I unfortunately cannot verify this.

I will continue working on performance and scalability and hope it will work as it should.

I will let the APIs stay as they are, the additional filters do not harm and can also help in other debug situations. I will close the issue as soon as I released the new version with all fixes.

**Comment by sherifkayad** [comment]
Sounds amazing! .. BTW I was thinking, maybe the query param filters are also useful for something like nested queries (or nested drop-downs) .. basically I select a certain namespace or a certain report type, severity, .. etc., I could get only the relevant `app` or `whatever` labels in that namespace matching that (those) report types or severity (severities) I selected.

Maybe a new ticket for a new future use case? what do you think?

**Comment by fjogeleit** [comment]
You mean the label values depending on the values of other filters? Would be possible for namespaces other values are related to results, so I would need to join them again.

**Comment by sherifkayad** [comment]
Yeah basically label values based on some other selected criteria .. if you have to join again and we get back to square 0 then forget what I said .. maybe that whole topic of filtering is a for another day .. as you said one could start with namespace filtering and then for other stuff, there could be another way in the future.

**Comment by fjogeleit** [comment]
Correct, also the rework of the complete UI is on the roadmap for this year. So any feedback is very appreciated.

**Comment by sherifkayad** [comment]
Awesome .. in that case I will then open a new ticket and maybe you can put it under the rework planned if it fits. I will try to express some "wishes" that I see with regards to filtering and will also mention that efforts there need to respect the query performance and avoid joins as much as possible. It won't be urgent by any means and you can start it whenever suitable.

**Comment by fjogeleit** [comment]
Sure, as you like. The "main" ticket is https://github.com/kyverno/policy-reporter/issues/222

The rework can also include restructering / reorganization of information. For example I think about organize results based on source rather than the type of Report.

So you have a page per source and don't jump between PolicyReport and ClusterPolicyReport pages.

**Comment by sherifkayad** [comment]
Maybe I put the detailed idea as a comment there or so. Will figure something out tomorrow or in the next few days.

**Comment by fjogeleit** [comment]
Take your time. I will also do some research before starting this topic

**Comment by fjogeleit** [comment]
Released with Helm Chart Version 2.15.0
